File: time_series/exercises/week7/arima_ex.R
# ----------------------------------------------------------------
# Example, investigate if arima() does fit a transfer function

# Generate a step input
n <- 500
x <- c(rep(0,100), rep(1,n-100))
# Add a step
plot(x, type="s")

# Make an output vector "filtered" by the system: y_t = 0.98 * y_t + x_t
y <- filter(x,0.98,"recursive")
# Add AR noise
y <- y + c(arima.sim(list(ar=c(0.7,0,0)), n))
# Plot the output
plot(y)

# Try to model with arima
m1 <- arima(y, xreg=x, order=c(2,0,0))
m1
tsdiag(m1)

# Predict to see the predicted one-step output
pred1 <- predict(m1, newxreg=x, n.ahead = 1)
# plot(pred1<span>$pred)
# There is no exponential response, so xreg is not included as a transfer function!

# Try to recreate the arima() with xreg is hard!
# Just LS directly, get parameters very different
lm(y ~ x)

# Conclusion:
# arima() does IN ONE GO: a linear model with xreg together with ARMA on residuals, 
# like e.g. ML estimation of ARMA(1,1) with one xreg: 
#   
#     (Y_t - beta_0 + x_t beta_1) = phi1 * (Y_{t-1} - beta_0 + x_{t-1} beta_1) + theta_0 eps_t + theta_1 eps_{t-1}
# 
# The parameters are estimated to minimize the squared residuals using an optimizer (so very close to maximum likelihood with normal distributed error)
# ----------------------------------------------------------------


# ----------------------------------------------------------------
# (1) ARX model model for a transfer function
# Make AR lags
y.l1 <- c(0,y[-n])
y.l2 <- c(0,y.l1[-n])

# Fit and one-step predictions with ARX(1)
arx1 <- lm(y ~ y.l1 + x)
summary(arx1)
plot(x, type="s")
plot(predict(arx1), type="l")
# ----------------------------------------------------------------


# ----------------------------------------------------------------
# Simulating some AR noise
set.seed(2)
noise <- arima.sim(model=list(ar=0.9), n = 300)
plot(noise)
# ----------------------------------------------------------------


# ----------------------------------------------------------------
# Harmonic oscillation + noise

# (2-3) Deterministic part with an amplitude:
amplitude <- 5
det <- amplitude * sin(0.1*(1:300)+1) + 2 
signal <- det + noise
plot(signal)
acf(signal) # Not stationary
acf(diff(signal)) # Looks like something odd creating a little correlation but no clear structure
(a1 <- arima(signal, order=c(0,1,0)))
# Validates?
tsdiag(a1) 
pacf(residuals(a1))
plot(residuals(a1))

# (4) Other model for amplitude 10
(a2 <- arima(signal, order=c(?,?,?)))
tsdiag(a2) # Looks good 

# (5) Let us try to predict with this model
plot(signal, xlim=c(0,400))
a2.pred <- predict(a2, n.ahead = 100)
matlines(301:400, a2.pred$</span>pred + 1.96*cbind(0, -a2.pred<span>$se, a2.pred$</span>se), col=2, lty=c(1,2,2))

# (6) Let us try a model using xreg with sin and cos with the true period
x <- cbind(sin = sin(0.1*(1:300)), cos = cos(0.1*(1:300)))
# We can fit a simple LM model:
lm1 <- lm(signal~x)
qqnorm(residuals(lm1)) # Looks fine
qqline(residuals(lm1))
summary(lm1) # All significant
acf(residuals(lm1)) # Ups - we should not trust the LM as residuals are correlated
pacf(residuals(lm1))
# Looks like AR(1)

# Using the same predictors and AR(1) structure for residuals
(a3 <- arima(signal, order=c(1,0,0), xreg = x, include.mean = TRUE))
tsdiag(a3) # Looks good

# Predict with harmonics model
plot(signal, xlim=c(0,400))
# We have to provide the xreg for the future
xpred <- cbind(sin = sin(0.1*(301:400)), cos = cos(0.1*(301:400)))
a3.pred <- predict(a3, n.ahead = 100, newxreg = xpred)
matlines(301:400, a2.pred<span>$pred + 1.96*cbind(0, -a2.pred$</span>se, a2.pred<span>$se), col=2, lty=c(1,2,2))
matlines(301:400, a3.pred$</span>pred + 1.96*cbind(0, -a3.pred<span>$se, a3.pred$</span>se), col=3, lty=c(1,2,2))
legend("bottomleft", legend = c("Using diff", "Using harmonic xreg"), lty=1, col=2:3, bg="white")
# ----------------------------------------------------------------


# ----------------------------------------------------------------
# Example: Linear trend + noise, using Arima from forecast package which automatically includes a slope
set.seed(2)
noise <- arima.sim(model=list(ar=0.9, ma=0.6), n = 300)
plot(noise)

det <- 0.05*(1:300+1) + 2
signal <- det + noise
plot(signal)
acf(signal) # Non stationary
acf(diff(signal)) 
pacf(diff(signal)) # More in PACF than ACF so maybe ARIMA(0,1,1)

a1 <- arima(signal, order=c(0,1,1))
tsdiag(a1) # Looks good
pacf(residuals(a1)) # Looks good
a1 # Note that you only estimate the ma1 parameter. There is no mean nor slope

# Predicting
plot(signal, xlim=c(0,400), ylim=c(-5,40))
# We have to provide the xreg for the future
a1.pred <- predict(a1, n.ahead = 100)
matlines(301:400, a1.pred<span>$pred + 1.96*cbind(0, -a1.pred$</span>se, a1.pred<span>$se), col=2, lty=c(1,2,2))
legend("bottomleft", legend = c("ARIMA(0,1,1)"), lty=1, col=2)
#Note that the prediction is a constant.
# 'arima' cannot estimate the slope of a trend without specifying it 

# The 'Arima' function from the forecast library can 'include.drift' (a slope)
library(forecast)
?Arima 
Arima # If you dig in you will see that it creates xreg ... see at the end

# Restarting with a simple linear regression (Same as lm(y~time))
a3 <- Arima(signal, order=c(0,0,0), include.drift = TRUE)
tsdiag(a3) # ACF is exponential
pacf(residuals(a3)) # Let us try an AR(1)

a4 <- Arima(signal, order=c(1,0,0), include.drift = TRUE)
tsdiag(a4) # Better but still something in lag 1
pacf(residuals(a4)) # Both lags 1 and 2 so try adding MA(1)

a5 <- Arima(signal, order=c(1,0,1), include.drift = TRUE)
tsdiag(a5) # Looks good
pacf(residuals(a5)) # Also good

# Fitting the same model using 'xreg' in 'arima'
a5b <- arima(signal, order=c(1,0,1), xreg = cbind(inter = 1, trend = 1:length(signal)), include.mean = FALSE)

# But let us try to predict with this model
plot(signal, xlim=c(0,400), ylim=c(-5,40))
# We have to provide the xreg for the future
lm.pred <- cbind(inter = 1, trend = (301:400))
a5b.pred <- predict(a5b, n.ahead = 100, newxreg = lm.pred)
matlines(301:400, a1.pred$</span>pred + 1.96*cbind(0, -a1.pred<span>$se, a1.pred$</span>se), col=2, lty=c(1,2,2))
matlines(301:400, a5b.pred<span>$pred + 1.96*cbind(0, -a5b.pred$</span>se, a5b.pred<span>$se), col=3, lty=c(1,2,2))
legend("topleft", legend = c("Using diff", "Using linear xreg"), lty=1, col=2:3)

File: time_series/exercises/week7/arima_ex.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.tsa.arima_process import arma_generate_sample
from statsmodels.tsa.statespace.sarimax import SARIMAX
from scipy.signal import lfilter

# Set seed for reproducibility
np.random.seed(2)

# ----------------------------------------------------------------
# 1. Transfer Function Investigation
# ----------------------------------------------------------------
n = 500
x = np.concatenate([np.zeros(100), np.ones(n - 100)])

# y_t = 0.98 * y_{t-1} + x_t
y_pure = lfilter([1.0], [1.0, -0.98], x)
# Add AR(1) noise: phi = 0.7
ar_params = np.array([1, -0.7])
ma_params = np.array([1])
noise = arma_generate_sample(ar_params, ma_params, n)
y = y_pure + noise

# Fit ARIMA with xreg (SARIMAX in Python)
# order=(p, d, q)
m1 = SARIMAX(y, exog=x, order=(2, 0, 0)).fit(disp=False)
print(m1.summary())

# Predict
# For one-step ahead with original xreg
pred1 = m1.get_prediction(exog=x).predicted_mean

# ----------------------------------------------------------------
# 2. Harmonic Oscillation + Noise
# ----------------------------------------------------------------
n_harm = 300
t = np.arange(1, n_harm + 1)
noise_harm = arma_generate_sample([1, -0.9], [1], n_harm)

# Signal: Amplitude 5 * sin(...) + 2
amplitude = 5
det = amplitude * np.sin(0.1 * t + 1) + 2
signal = det + noise_harm

# Harmonic Regressors (xreg)
exog_harm = pd.DataFrame({
    'sin': np.sin(0.1 * t),
    'cos': np.cos(0.1 * t)
})

# Fit AR(1) with Harmonic xreg
# include_intercept=True is default via 'c' in trend or just having the mean in SARIMAX
a3 = SARIMAX(signal, exog=exog_harm, order=(1, 0, 0), trend='c').fit(disp=False)

# Forecasting 100 steps ahead
t_future = np.arange(301, 401)
exog_forecast = pd.DataFrame({
    'sin': np.sin(0.1 * t_future),
    'cos': np.cos(0.1 * t_future)
}, index=t_future)

forecast_res = a3.get_forecast(steps=100, exog=exog_forecast)
mean_forecast = forecast_res.predicted_mean
conf_int = forecast_res.conf_int(alpha=0.05) # 95% CI

# Plotting
plt.figure(figsize=(10, 5))
plt.plot(t, signal, label='Observed')
plt.plot(t_future, mean_forecast, color='green', label='Harmonic xreg Forecast')
plt.fill_between(t_future, conf_int.iloc[:, 0], conf_int.iloc[:, 1], color='green', alpha=0.2)
plt.title("Harmonic Oscillation Forecast")
plt.legend()
plt.show()

# ----------------------------------------------------------------
# 3. Linear Trend (Drift) + Noise
# ----------------------------------------------------------------
n_trend = 300
t_trend = np.arange(1, n_trend + 1)
noise_trend = arma_generate_sample([1, -0.9], [1, 0.6], n_trend)
signal_trend = (0.05 * t_trend + 2) + noise_trend

# Python SARIMAX handles drift via the 'trend' parameter
# 'c' = intercept, 't' = time trend, 'ct' = both
a5 = SARIMAX(signal_trend, order=(1, 0, 1), trend='ct').fit(disp=False)
print(a5.summary())

# Predict 100 steps
forecast_trend = a5.get_forecast(steps=100)
mean_t_forecast = forecast_trend.predicted_mean

plt.figure(figsize=(10, 5))
plt.plot(t_trend, signal_trend, label='Data')
plt.plot(np.arange(301, 401), mean_t_forecast, label='Forecast with Trend', color='red')
plt.legend()
plt.show()